api/src/oscc.c

```diff
@@ -492,110 +614,456 @@ oscc_result_t oscc_async_enable( int socket )
+    can_contains_s detection =
+    {
+        .is_oscc = oscc_detection.has_brake_report &&
+                   oscc_detection.has_steer_report &&
+                   oscc_detection.has_accel_report,
+        .has_vehicle = vehicle_detection.has_brake_pressure &&
+                       vehicle_detection.has_steering_angle &&
+                       vehicle_detection.has_wheel_speed
+    };
+
+    return detection;
+}
+
+
+oscc_result_t construct_interfaces_list( device_names_s * const names_ptr )
+{
+    FILE *file_handler;
+    char buffer[512];
+    oscc_result_t result = OSCC_OK;
+
+    file_handler = fopen( "/proc/net/dev", "r" );
+    if (!file_handler) {
+        perror( "Cannot read: /proc/net/dev" );
+
+        result = OSCC_ERROR;
+    }
+
+    if( result == OSCC_OK && names_ptr == NULL )
     {
-        ret = oscc_async_enable( sock );
+        result = OSCC_ERROR;
+    }
+
+    int lines = 0;
+
+    if( result == OSCC_OK )
+    {
+        while ( fgets( buffer, sizeof( buffer ), file_handler ) ) {
+            ++lines;
+        }
+
+        names_ptr->name = (char**) malloc(lines * sizeof(char*));
 
-        if ( ret != OSCC_OK )
+        uint i;
+
+        for ( i = 0; i < lines; i++ )
         {
-            printf( "Enabling asynchronous socket I/O failed\n" );
+            names_ptr->name[i] = (char*) malloc(IFNAMSIZ * sizeof(char));
+        }
 
-            result = OSCC_ERROR;
+        rewind( file_handler );
+
+        //Consume the first two lines since they are headers
+        fgets( buffer, sizeof( buffer ), file_handler );
+        fgets( buffer, sizeof( buffer ), file_handler );
+    }
+
+    char* socket_name = calloc( IFNAMSIZ, sizeof(char) );
+
+    if( !socket_name )
+    {
+        result = OSCC_ERROR;
+    }
+
+    if( result == OSCC_OK )
+    {
+        uint size = 0;
+
+        while ( size < lines && fgets( buffer, sizeof( buffer ), file_handler ) ) {
+
+            result = get_device_name( buffer, socket_name );
+
+            if( result == OSCC_OK )
+            {
+                strncpy( names_ptr->name[size], socket_name, IFNAMSIZ );
+
+                size++;
+            }
         }
+
+        free( socket_name );
+
+        names_ptr->size = size;
     }
 
+    fclose( file_handler );
 
-    if ( result == OSCC_OK )
+    return result;
+}
+
+oscc_result_t clear_device_names( device_names_s * const names_ptr )
+{
+    oscc_result_t result = OSCC_OK;
+
+    if( names_ptr == NULL )
     {
-        /* all prior checks will pass even if a valid interface has not been
-           set up - attempt to write an empty CAN frame to the interface to see
-           if it is valid */
-        struct can_frame tx_frame;
+        result = OSCC_ERROR;
+    }
+    else
+    {
+        uint i;
 
-        memset( &tx_frame, 0, sizeof(tx_frame) );
-        tx_frame.can_id = 0;
-        tx_frame.can_dlc = 8;
+        for ( i = 0; i < names_ptr->size; i++ )
+        {
+            free( names_ptr->name[i] );
```

Comment:
Check for NULL before `free` then set to pointer to NULL here and below. Calling this twice should be a no-op rather than a segfault.